modular_citadel/code/modules/nifsoft/nif.dm

```diff
@@ -1,8 +1,857 @@
+/obj/item/device/nif/proc/deactivate_these(var/list/turn_off)
+	for(var/N in turn_off)
+		var/datum/nifsoft/NS = nifsofts[N]
+		if(NS)
+			NS.deactivate()
+
+//Add a flag to one of the holders
+/obj/item/device/nif/proc/set_flag(var/flag,var/hint)
+	ASSERT(flag != null && hint)
+
+	switch(hint)
+		if(NIF_FLAGS_VISION)
+			vision_flags |= flag
+		if(NIF_FLAGS_HEALTH)
+			health_flags |= flag
+		if(NIF_FLAGS_COMBAT)
+			combat_flags |= flag
+		if(NIF_FLAGS_OTHER)
+			other_flags |= flag
+		else
+			CRASH("Not a valid NIF set_flag hint: [hint]")
+
+//Clear a flag from one of the holders
+/obj/item/device/nif/proc/clear_flag(var/flag,var/hint)
+	ASSERT(flag != null && hint)
+
+	switch(hint)
+		if(NIF_FLAGS_VISION)
+			vision_flags &= ~flag
+		if(NIF_FLAGS_HEALTH)
+			health_flags &= ~flag
+		if(NIF_FLAGS_COMBAT)
+			combat_flags &= ~flag
+		if(NIF_FLAGS_OTHER)
+			other_flags &= ~flag
+		else
+			CRASH("Not a valid NIF clear_flag hint: [hint]")
+
+//Check for an installed implant
+/obj/item/device/nif/proc/imp_check(var/soft)
+	if(stat != NIF_WORKING) return FALSE
+	ASSERT(soft)
+
+	if(ispath(soft))
+		var/datum/nifsoft/path = soft
+		soft = initial(path.list_pos)
+	var/entry = nifsofts[soft]
+	if(entry)
+		return entry
+
+//Check for a set flag
+/obj/item/device/nif/proc/flag_check(var/flag,var/hint)
+	if(stat != NIF_WORKING) return FALSE
+
+	ASSERT(flag && hint)
+
+	var/result = FALSE
+	switch(hint)
+		if(NIF_FLAGS_VISION)
+			if(flag & vision_flags) result = TRUE
+		if(NIF_FLAGS_HEALTH)
+			if(flag & health_flags) result = TRUE
+		if(NIF_FLAGS_COMBAT)
+			if(flag & combat_flags) result = TRUE
+		if(NIF_FLAGS_OTHER)
+			if(flag & other_flags) result = TRUE
+		else
+			CRASH("Not a valid NIF flag hint: [hint]")
+
+	return result
+
+/obj/item/device/nif/proc/planes_visible()
+	if(stat != NIF_WORKING)
+		return list() //None!
+
+	return planes_visible
+
+/obj/item/device/nif/proc/add_plane(var/planeid = null)
+	if(!planeid)
+		return
+	planes_visible |= planeid
+
+/obj/item/device/nif/proc/del_plane(var/planeid = null)
+	if(!planeid)
+		return
+	planes_visible -= planeid
+
+/obj/item/device/nif/proc/vis_update()
+	if(human)
+		human.recalculate_vis()
+
+// Alternate NIFs
+/obj/item/device/nif/bad
+	name = "bootleg NIF"
+	desc = "A copy of a ripoff of a copy of a clone of... this can't be any good, right?"
+	typeletter = "B"
+	quality = -1 //Ignores quality restrictions
+	durability = 30
+	bioadap = TRUE
+	charge_use_multiplier = 0.1
+
+	emag_act(mob/user)
+		if(open)
+			var/obj/item/device/nif/newnif = new /obj/item/device/nif/sketchy()
+			newnif.owner = owner
+			newnif.open = open
+			newnif.update_icon()
+
+			newnif.forceMove(get_turf(src))
+			QDEL_NULL(src)
+		. = ..()
+
+/obj/item/device/nif/sketchy //Ignores NIFSoft Vendor Access Restrictions, handled in the vendor. Not a free emag.
+	name = "sketchy NIF"
+	stealth_name = "bootleg NIF"
+	desc = "An illegally modified NIF which is definitely major contraband, and malpractice to install."
```

Comment:
Don't hard-code legality of items. Send this to the admin team for the wiki.

Reply:
It can install the known illegal nifsofts from a hacked nifsoft vendor.